File: go/main_test.go
package main_test

import (
	"bytes"
	"errors"
	"testing"

	sub "github.com/Heliodex/substitution"
)

// Test multiple substitutions of the same name
var s = sub.Sub{
	PartNames: []sub.PartName{
		{Part: "Hello ", Name: "name"},
		{Part: "! You have ", Name: "count"},
		{Part: " new messages. Thanks, ", Name: "name"},
	},
	Final: "!",
}

func TestSubstitute(t *testing.T) {
	toSub := sub.ToSub{
		"name":  "Heliodex",
		"count": "67",
	}

	result, err := s.Sub(toSub)
	if err != nil {
		t.Fatal(err)
	}

	if result != "Hello Heliodex! You have 67 new messages. Thanks, Heliodex!" {
		t.Fatalf("unexpected result: %s", result)
	}
}

func TestSubstituteMissing(t *testing.T) {
	toSub := sub.ToSub{
		"name": "Heliodex",
	}

	result, err := s.Sub(toSub)
	if err == nil {
		t.Fatalf("expected error, got result: %s", result)
	}
	if !errors.Is(err, sub.ErrMissingValue) {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestSubstituteExtra(t *testing.T) {
	toSub := sub.ToSub{
		"name":  "Heliodex",
		"count": "67",
		"extra": "value",
	}

	result, err := s.Sub(toSub)
	if err == nil {
		t.Fatalf("expected error, got result: %s", result)
	}
	if !errors.Is(err, sub.ErrExtraValue) {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestSerialisation(t *testing.T) {
	data := s.Serialise()

	s2, err := sub.DeserialiseSub(bytes.NewReader(data))
	if err != nil {
		t.Fatal(err)
	}

	if !s.Equals(s2) {
		t.Fatalf("deserialisation mismatch: expected %q, got %q", s.Final, s2.Final)
	}
}

func TestDeserialisationTooShortLength(t *testing.T) {
	s, err := sub.DeserialiseSub(bytes.NewReader([]byte{0, 0, 0}))
	if err == nil {
		t.Fatalf("expected error, got s: %v", s)
	}
	if !errors.Is(err, sub.ErrDataTooShortLength) {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestDeserialisationTooShortString(t *testing.T) {
	_, err := sub.DeserialiseSub(bytes.NewReader([]byte{0, 0, 0, 1, 0, 0, 0, 1}))
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	if !errors.Is(err, sub.ErrDataTooShortString) {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestEquals(t *testing.T) {
	s2 := sub.Sub{
		PartNames: []sub.PartName{
			{Part: "Hello ", Name: "name"},
			{Part: "! You have ", Name: "count"},
			{Part: " new messages. Thanks, ", Name: "name"},
		},
		Final: "!",
	}

	if !s.Equals(s2) {
		t.Fatal("expected subs to be equal")
	}

	s3 := sub.Sub{
		PartNames: []sub.PartName{
			{Part: "Sup ", Name: "name"},
			{Part: ", you got ", Name: "num"},
			{Part: " new pings. Thanks, ", Name: "name"},
		},
		Final: "!",
	}

	if s.Equals(s3) {
		t.Fatal("expected subs to not be equal")
	}
}

File: ts/main.test.ts
import { expect, test } from "bun:test"
import * as sub from "./main"

// Test multiple substitutions of the same name
const s = new sub.Sub(
	[
		new sub.PartName("Hello ", "name"),
		new sub.PartName("! You have ", "count"),
		new sub.PartName(" new messages. Thanks, ", "name"),
	],
	"!"
)

test("Substitute", () => {
	const toSub = { name: "Heliodex", count: "67" } as sub.ToSub
	const result = s.Sub({ ...toSub })
	expect(result).toBe("Hello Heliodex! You have 67 new messages. Thanks, Heliodex!")
})

test("SubstituteMissing", () => {
	const toSub = { name: "Heliodex" } as sub.ToSub
	let err: Error | null = null
	try {
		s.Sub({ ...toSub })
	} catch (e) {
		err = e as Error
	}
	expect(err).not.toBeNull()
	expect(err?.message).toContain(sub.ErrMissingValue.message)
})

test("SubstituteExtra", () => {
	const toSub = { name: "Heliodex", count: "67", extra: "value" } as sub.ToSub
	let err: Error | null = null
	try {
		s.Sub({ ...toSub })
	} catch (e) {
		err = e as Error
	}
	expect(err).not.toBeNull()
	expect(err?.message).toContain(sub.ErrExtraValue.message)
})

test("Serialisation", () => {
	const data = s.Serialise()
	const s2 = sub.Sub.Deserialise(Buffer.from(data))
	expect(s.Equals(s2)).toBe(true)
})

test("DeserialisationTooShortLength", () => {
	let err: Error | null = null
	try {
		sub.Sub.Deserialise(Buffer.from([0, 0, 0]))
	} catch (e) {
		err = e as Error
	}
	expect(err).not.toBeNull()
	expect(err).toBe(sub.ErrDataTooShortLength)
})

test("DeserialisationTooShortString", () => {
	let err: Error | null = null
	try {
		sub.Sub.Deserialise(Buffer.from([0, 0, 0, 1, 0, 0, 0, 1]))
	} catch (e) {
		err = e as Error
	}
	expect(err).not.toBeNull()
	expect(err).toBe(sub.ErrDataTooShortString)
})

test("Equals", () => {
	const s2 = new sub.Sub(
		[
			new sub.PartName("Hello ", "name"),
			new sub.PartName("! You have ", "count"),
			new sub.PartName(" new messages. Thanks, ", "name"),
		],
		"!"
	)

	expect(s.Equals(s2)).toBe(true)

	const s3 = new sub.Sub(
		[
			new sub.PartName("Sup ", "name"),
			new sub.PartName(", you got ", "num"),
			new sub.PartName(" new pings. Thanks, ", "name"),
		],
		"!"
	)

	expect(s.Equals(s3)).toBe(false)
})
